### PR TITLE
diglib: Calculate the total size of topo file to get the correct off_t_size

### PR DESCRIPTION
**Pull request by HuidaeCho**
Creating a new PR for #3163 (accidentally merged instead of rebase).

`off_t` sizes for coor and topo files can be different. For example, from an EPSG:4326 location,
```bash
> v.mkgrid map=grids_00625deg position=coor coor=-180,-90 box=0.0625,0.0625 grid=2880,5760
> d.mon wx0
> d.vect grids_00625deg
ERROR: Unable to seek: Invalid argument

> ls -al .../vector/grids_00625deg
-rw-r--r-- 1 hcho users 1708966098 Sep 17 22:35 coor
-rw-r--r-- 1 hcho users         95 Sep 17 22:33 dbln
-rw-r--r-- 1 hcho users        173 Sep 17 22:33 head
-rw-r--r-- 1 hcho users        260 Sep 17 22:33 hist
-rw-r--r-- 1 hcho users 6706995200 Sep 17 22:45 sidx
-rw-r--r-- 1 hcho users 2173694570 Sep 17 22:45 topo
```
`coor_size` is less than `PORT_LONG_MAX` (2^31 - 1 = 2147483647; max `int`), but the topo size is greater than that. Setting `ptr->off_t_size = 4` in https://github.com/OSGeo/grass/blob/fd59e94985870d8558e236cf8015a99b8d83e5ff/lib/vector/diglib/plus_struct.c#L677-L682 causes an integer overflow and `ptr->Isle_offset` is read as -2121341854 in https://github.com/OSGeo/grass/blob/fd59e94985870d8558e236cf8015a99b8d83e5ff/lib/vector/diglib/plus_struct.c#L641

This PR calculates the total size of topo file to get the correct `off_t_size`. Before this PR, the size of coor file was used to determine this offset size for topo file.

### CODE REVIEW

File: lib/vector/diglib/plus_struct.c [hunk 1]
```diff
@@ -678,8 +692,69 @@ int dig_Wr_Plus_head(struct gvfile *fp, struct Plus_head *ptr)
...
+            /* from dig_Wr_P_isle() */
+            struct P_isle *p = ptr->Isle[i];
+
+            if (p == NULL)
+                size += 4;
+            else
+                size += 8 + 4 * p->n_lines;
+        }
+
+        ptr->off_t_size = size > PORT_LONG_MAX ? 8 : 4;
```
Comment (metzm) [inline lib/vector/diglib/plus_struct.c:756]: `PORT_LONG_MAX` is an int, cast to off_t for correct comparison as above with `if (ptr->coor_size > (off_t)PORT_LONG_MAX) {`
```suggestion
        ptr->off_t_size = size > (off_t)PORT_LONG_MAX ? 8 : 4;
```

File: lib/vector/diglib/plus_struct.c [hunk 2]
```diff
@@ -678,8 +692,69 @@ int dig_Wr_Plus_head(struct gvfile *fp, struct Plus_head *ptr)
         /* can only happen when sizeof(off_t) == 8 */
         ptr->off_t_size = 8;
     }
-    else
-        ptr->off_t_size = 4;
+    else if (ptr->off_t_size == 0) {
```
Comment (metzm) [inline lib/vector/diglib/plus_struct.c:695]: This test is here to skip the second calculation of the topo file size when `dig_Wr_Plus_head()` is called in `dig_write_plus_file()` the second time?

Comment (HuidaeCho) [reply lib/vector/diglib/plus_struct.c:695]: Yes, you're right. `dig_Wr_Plus_head()` gets called twice from `dig_write_plus_file()` and the topo size calculation is rather expensive.

**Review by metzm — APPROVED** [review]
Thanks for this bugfix!

**Comment by echoix** [comment]
This PR is approved and reviewed. Is there anything that should block merging this?

**Pull request merged**